### DISCUSSION TITLE
对X-cmd的理解

### DISCUSSION THREAD

**Discussion by MaxforCherubim** [x-cmd/x-cmd#67] [Q&A] [answered]
统计学出身刚学习计算机1年有余的小白。
无意中在B站上刷到了x-cmd，发现似乎是个工具集合包，其中包括了部分我在使用的fzf、bat、tldr等命令行工具。
本来看到付费咨询服务，特地想打钱给你们并咨询学习一下😊，不过发现还没开通是吧？
目前萦绕在我脑海中最大的问题就是，我mac已经通过brew安装了相当部分的命令行工具，还特地模仿他人的.zshrc自定义自己的。
如果我安装了x-cmd并使用它的比如说tldr，是会调用之前brew已经安装好的tldr吗？
我已经看到文档里说是根据.zshrc中的读取顺序加载的，那么有没有必要卸载已经安装的命令行工具，毕竟x-cmd都已经提供了。
主要是因为我有点完美主义的感觉，就那种没用的东西为什么不卸载算了。

**Comment by edwinjhlee** [comment]
你好。

x-cmd 的 模块 调用的是 同名命令，还是自行实现，这个要看情况。

## `x tldr` 是 x-cmd 自行采用 shell 实现的，并未采用 其它 二进制 client

以 tldr 模块为例，我们 采用的是 x-cmd 采用 shell 和 curl 实现的 方案，而不是 官方的 tldr client。

我们是出于这些原因考虑：

1.  tldr 之前的 client 都比较重，而最近的 rust 版本也是去年才提供。而我们一旦采用 python 和 node 版本，意味着 tldr 模块 要先给用户准备好 node （20MB+） 或者 python （30MB+） 环境才能使用。x-cmd 一直追求轻量化、透明化，因此我们就自行编写了 tldr client -- x-cmd 压缩代码包才1.2MB，tldr 英文的压缩包也只有 600KB。加起来不到 2MB，何必要 20MB ？
2. 官方 tldr 的 client 很多，但似乎都不愿意在 交互上 进一步深入。我们的 [tldr](https://cn.x-cmd.com/mod/tldr) 应该是 当前 交互性最强的 tldr client 。我们后面 还会引入 fzf 进行 进一步增强。

## x-cmd 自带了 `pkg` ，与系统 pkg 管理器 和 其它 第三方管理器 （例如 pixi，pkgx ） 并非是 竞争，而是 互补关系

x-cmd 的 pkg 管理有自行的设计原则。x-cmd 大部分的 软件要么是静态编译，要么都是自带依赖的。因此在安装时不会像其它包管理器有大量的依赖计算和分别下载。在很多时候，x-cmd 的 pkg 会更小，下载也会更快。但其它的包管理器当前的实现也有别的优势（例如，共享相同的依赖从而减少硬盘占用，等等）。

但由于我们的精力有限，x-cmd 的 包管理体系 所支持的 包会相对更少：

1. 我们不愿意花费精力去提供一些 不维护的软件
2. 别的 pkg manager 能做好的，我们也尽量不做。例如 pixi 背后的 conda-forge 体系 就很强大，我们就尽量不重复实现。

因此，x-cmd 支持一系列的第三方包管理器。有 pixi，asdf，pkgx
https://cn.x-cmd.com/mod/pixi 

另外，x-cmd 也对 apt，yum，brew（其实也是第三方） 这类 系统级别的 包管理器 提供了 增强。

我们会尝试把同一软件的多种安装方式在 `x install` 中提供。https://cn.x-cmd.com/install


## x-cmd 模块 是按需下载 并调用其它二进制工具，不需要 root，也力求不影响 用户当前环境（出于 安全性，系统环境的稳定性）

以 `x jq` 为例，x-cmd：

-  先看看当前环境 有没有 jq，
   - 有的话，就直接使用；
   - 如果没有，就看是否之前 pkg 是否下载并解压：
      - 如果没有，就下载并解压，并调用 jq
      - 如果有，那么就用 pkg 之前下载并解压的 jq
      
注意，x-cmd 此时仍并不会安装，意味着你仍不能在环境中直接使用 `jq`，但你能够通过 `x jq` 调用。
如果你需要直接 使用 `jq`，采用 `x env use jq` 或者 `x env try jq` 。


## x-cmd 尝试友好，但尽量避免做简化模型

因为，x-cmd 的设计考虑到 ：

1. 服务器运维 要安全、规范 （ 因此，我们要尽量减少对 环境的影响 ）
2. 开发人员 要 方便地隔离环境，安装最新的开发 sdk

我们尽量做得友好，尤其对小白。但我们无意于简化 -- 因为简化意味着我们不得不在事物自然的模型 A 中提炼一个看似简化的模型 B， 以及 一系列新造的 术语。而 小白/新手 只是一个短暂的阶段，等用户 熟练之后，就不得不 为 多余的 模型 B 以及 一系列 新造的术语 而增加 更多的 认知负担，从而 可能 带来 更多的 混淆和失误。


## 择一

我很欣赏这种思路。不过，由于我是开发者，一方面 不会 公开 dis 第三方工具，另一方面，我在开发过程中也试用了相当多的工具、系统，除了少数项目因为年久失修而失去价值，大部分 项目 都有些适用场景的亮点。

我建议考虑 和我 持有一样的 心态，就是 将 x-cmd 视为探索的工具。

以 [`x install`](https://x-cmd.com/install) 为例，我们尝试给用户提供更多的安装选择，用户可以根据场景，考虑是否采用最新版本，是否全局安装，再自行通过 TUI 选择安装命令。虽然比 `x <bin>` 要多一个选择，但整个过程也是相当轻松，我们还会尽量在日后加入 各个不同安装方式的 背后的仓库地址，以及 跳转命令，还有 介绍。

又例如，我们最近 仅 top，就提供 [btop](https://x-cmd.com/mod/btop), [htop](https://x-cmd.com/mod/htop)， bottom 等等。


不过，我们还是会做减法的。一些不维护的软件包，我们的 `pkg` 虽然会保持下载源，但不会在 补全 和 文档中提及。这其实也是 从 减少 用户认识负担 角度出发。


## x-cmd 不仅仅是 pkg

x-cmd 本质是一个 shell 的代码库，在此之上实现了一系列命令行工具，也增强 了一系列的 命令行工具。

正因为 部分的 命令行工具 不会在 系统预装，我们才引入了 `pkg` 体系。

**Reply by WayenVan** [reply]
可以说是新时代的busybox 点赞

**Comment by MaxforCherubim** [comment]
感谢回复！
我现在感觉x-cmd就是一个集成工具，我现在在wsl2上使用它，这下不用花费大半天甚至一整天的时间去把原来mac上的配置完全照搬过来了，很够用了！
不过还有个问题，mac一般是用brew安装工具的，后续如果用x brew安装工具时是会安装到x brew 自己的路径吗？
感觉大部分方便的工具都有，看起来就没有x-cmd做不到的，以后新电脑直接安装一个x-cmd是否能代替git、brew等一系列常用工具呢？都不用刻意去装git、brew等？

**Comment by edwinjhlee** [comment]
不是， x brew 就是 调用 系统的 brew （ 所以 跟你直接 调用 brew 是一样的，如果不一样，那就是 bug，欢迎 发 issue ），只是我们在上面做了一些 UI 增强。例如 浏览 brew 上的 包 。 

不过，x-cmd 本身并不仅仅是 集成，只不过 shell 本身非常适合 作为 各种工具的 黏结 。这就是 为啥我放弃了之前 采用 typescript，go，rust 和 c 的 实现，采用 posix shell 这种非常小众的 工程化 很难 的语言。当然，这仅是其中一个原因。

---

我经常用  无限手套 来比喻 x-cmd 。brew 等一众工具 就是里面的 宝石 ； 而 仅有 宝石，而没有 无限手套，也很难 发挥出 这些 宝石 的联合起来 的巨大作用。

---

x-cmd 在最近两年已经成为我安装新电脑的第一个工具 ～ 很爽 ～ 尤其 在国内 也运作流畅。

不过，部分包最好还是采用 系统管理器 或者 系统默认的方式来安装使用。例如 git。一方面因为部分 应用  会默认 去系统目录找 git；或者 mac 直接将 git 放进其 开发者工具里面，而这个 git 应该和 苹果自身的开发工具链 结合得 更紧，出问题的概率也更小。

现阶段，x-cmd 更倾向于解决那些 小、新、更新迭代快 的 工具软件。例如，x-cmd 内部的 git 模块，会优先使用环境的 git 命令，但是如果 环境 没有，我们就会 用 jijistu （ jj 模块） ~  一个新的 git 兼容软件，体积更小因此下载更快，非常符合 x-cmd 的审美。

我仍然认为 git 是不可替代的，但这不妨碍我们尝试应用更有趣的新选项。

---

我明白 你 可能 希望 一个工具解决所有问题。但是，世界太大了，非常精采，有些领域水也很深；而 x-cmd 的团队太小了，见识也不广 ～ 不过呢，我们很乐意一点点探索世界，并把些攻略以合适的方式像马赛克一样镶嵌在 x-cmd 里面。

欢迎你多使用，如果 有合适的 应用，欢迎推荐。我们不一定会马上放在 x-cmd 里面，因为 x-cmd 也要考虑到 稳定、准确、以及日后的兼容问题。但我们会充分学习和考虑最佳的实现形态。

随着我们的 hub 在未来两个月内发布，我们和用户都可以 更灵活自由 探索更多的 与 x-cmd 集成和增强的 模式 ～～～

**Comment by MaxforCherubim** [comment]
我现在有点明白了。
x [mod] 是用shell和awk实现的命令，所以无需安装或者自动安装就可以直接运行 x [mod]命令，对应文档里的`模块`；
x env 对应的是包和依赖环境，这种的需要加载正确环境才能使用的包，对应文档里的`包`；
x install 更是使得用户可以自己安装需要的，同时让用户更加方便，无需去找官方正确的安装命令，对应文档里的`install`。
我这样理解正确吗？

**Comment by edwinjhlee** [comment]
> 我现在有点明白了。 x [mod] 是用shell和awk实现的命令，所以无需安装或者自动安装就可以直接运行 x [mod]命令，对应文档里的`模块`； x env 对应的是包和依赖环境，这种的需要加载正确环境才能使用的包，对应文档里的`包`； x install 更是使得用户可以自己安装需要的，同时让用户更加方便，无需去找官方正确的安装命令，对应`install`。 我这样理解正确吗？


基本没错。 x [mod] 也会帮你按需调用 pkg 来下载一些包 （ 但不会安装影响你当前 环境 ）。

你可以先按照这个理解开始运作，用得越多，往后可能会有更细致的理解。

我们也在构思如何把这件事 在 文档上 讲得更清楚。

---

准确来说，pkg 是包管理（下载，解包，环境加载）。而 env 只是 包管理的一种应用。我也很想对用户 屏蔽 这个细节。但 我们还有 开发者用户。有时候 需要 了解这个细节 进行 更精细的 调控。

**Comment by MaxforCherubim** [comment]
也就是说mac的brew、ubuntu的apt、python的conda等是pkg，我刚试用的x env jieba这个jieba就是env咯？因为我在使用这个jieba的时候，看到了安装过程中的miniconda，但是只能使用jieba命令，而无法使用conda命令。

**Comment by edwinjhlee** [comment]
> 也就是说mac的brew、ubuntu的apt、python的conda等是pkg，我刚试用的x env jieba这个jieba就是env咯？因为我在使用这个jieba的时候，看到了安装过程中的miniconda，但是只能使用jieba命令，而无法使用conda命令。

是的。

因为，如果我们把 jieba 所依赖的 python 版本（打个比方， 3.12） 随着 jieba 的安装 注入到 当前环境，如此就不经意 改变了 你当前环境的 python 了（例如 你 当前开发环境用的是 3.9 )。

如果 你想 显式 使用当前环境 python （ 3.9 ） 下 所安装 jieba，用 pipx 或者 直接 pip install 会 显式，也更好。

**Comment by lunrenyi** [comment]
@MaxforCherubim 
`x env ls -i` 命令：查看 x-cmd pkg 已安装的软件工具 （或者使用 `x pkg ls` 命令）
`x env ls` 命令：查看已在环境中激活使用的 x-cmd pkg 软件工具
`x env ll` 命令：查看 x-cmd pkg 支持的软件工具

**Comment by MaxforCherubim** [comment]
好家伙了，我又多了一个python，我似乎有点明白了unix系统以文件组织的道理了，直接在.x-cmd.root文件夹下。
我又使用了usql，但是我发现usql为什么在文档中是属于`包`的，却可以使用x usql，用模组形式的命令来使用，是因为这样可以联动其他模组吗？

**Comment by MaxforCherubim** [comment]
> @MaxforCherubim `x env ls -i` 命令：查看 x-cmd pkg 已安装的软件工具 （或者使用 `x pkg ls` 命令） `x env ls` 命令：查看已在环境中激活使用的 x-cmd pkg 软件工具 `x env ll` 命令：查看 x-cmd pkg 支持的软件工具

记住了，感谢回复！

**Comment by edwinjhlee** [comment]
> 好家伙了，我又多了一个python，我似乎有点明白了unix系统以文件组织的道理了，直接在.x-cmd.root文件夹下。 我又使用了usql，但是我发现usql为什么在文档中是属于`包`的，却可以使用x usql，用模组形式的命令来使用，是因为这样可以联动其他模组吗？

主要是，如果我们 某个工具 用了系统的 python，有一天 系统的 python 被卸载了，或者 升降级，那么 这个 工具 也可能受到影响。

放心，我们尽量地复用这个 自管理的 python，而且 当所有依赖它的 工具 被 卸载了，这个 python 也会随之 被 gc 。

**Comment by edwinjhlee** [comment]
> 好家伙了，我又多了一个python，我似乎有点明白了unix系统以文件组织的道理了，直接在.x-cmd.root文件夹下。 我又使用了usql，但是我发现usql为什么在文档中是属于`包`的，却可以使用x usql，用模组形式的命令来使用，是因为这样可以联动其他模组吗？

除非你 还有 其它 shell 或者 进程 调用 x-cmd，如果你确定 没有，或者 不会出现问题。有些时候 直接 `rm -rf ~/.x-cmd.root` 然后 再重装 x-cmd 是一种 很好故障恢复的 手段。

x-cmd 对系统的侵入性 基本都在 ~/.x-cmd.root 里面了。

---

`rm -rf` 大法有风险，使用 要 谨慎 ～

**Comment by MaxforCherubim** [comment]
> > 好家伙了，我又多了一个python，我似乎有点明白了unix系统以文件组织的道理了，直接在.x-cmd.root文件夹下。 我又使用了usql，但是我发现usql为什么在文档中是属于`包`的，却可以使用x usql，用模组形式的命令来使用，是因为这样可以联动其他模组吗？
> 
> 除非你 还有 其它 shell 或者 进程 调用 x-cmd，如果你确定 没有，或者 不会出现问题。有些时候 直接 `rm -rf ~/.x-cmd.root` 然后 再重装 x-cmd 是一种 很好故障恢复的 手段。
> 
> x-cmd 对系统的侵入性 基本都在 ~/.x-cmd.root 里面了。
> 
> `rm -rf` 大法有风险，使用 要 谨慎 ～

好的明白，任何学过linux都知道的。

**Comment by edwinjhlee** [comment]
usql 没有 相应的 module，我们 是通过一个 叫做 xbin 的机制。

我承认这件事有点乱。但也没有办法。因为一个 pkg 可能有很多 bin，如果 所有这些 bin 都是 module，那么 module 数目会 暴增，而且 pkg 之前 也有 重复的 bin。

你现在 可以在 x-cmd 的 搜索里面 看到 各个包 可用的方法 （ 有些 包我们还没有 来得及 介绍 ），能用 x <bin> 模式的，我们都在右上角标注了。

**Comment by Zhengqbbb** [comment]
![CleanShot 2024-08-23 at 18 12 54@2x](https://github.com/user-attachments/assets/10a8e44d-51c4-44bb-8e5d-8434534f0b80)

**Comment by MaxforCherubim** [comment]
> usql 没有 相应的 module，我们 是通过一个 叫做 xbin 的机制。
> 
> 我承认这件事有点乱。但也没有办法。因为一个 pkg 可能有很多 bin，如果 所有这些 bin 都是 module，那么 module 数目会 暴增，而且 pkg 之前 也有 重复的 bin。
> 
> 你现在 可以在 x-cmd 的 搜索里面 看到 各个包 可用的方法 （ 有些 包我们还没有 来得及 介绍 ），能用 x 模式的，我们都在右上角标注了。

嗯嗯，我就是看到有些pkg只有env use，有些却同时还有x [mod]。
等于说这种形式就是为了使得bin不重复，实际上x usql不属于真的x [mod]，是这样的吗？
不好意思，统计学出身，学数学学的这样，会想在脑海中形成一个范式。

**Comment by MaxforCherubim** [comment]
> ![CleanShot 2024-08-23 at 18 12 54@2x](https://private-user-images.githubusercontent.com/40693636/360890569-10a8e44d-51c4-44bb-8e5d-8434534f0b80.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjQ0MDgzMDksIm5iZiI6MTcyNDQwODAwOSwicGF0aCI6Ii80MDY5MzYzNi8zNjA4OTA1NjktMTBhOGU0NGQtNTFjNC00NGJiLThlNWQtODQzNDUzNGYwYjgwLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA4MjMlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwODIzVDEwMTMyOVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWNlYzNkYmM5N2M0MjgyMzQ0MTVhYjQ0ZDlkM2Y4Y2JmNDlhOTM4ZGM0NzUyYjllZjc3MjgxOWE5Y2E3MzQzMTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.6DO5UIoj4i9uChr63hQsnE7bqJvk8ajVY4lsdsBsFQ4)

收到，感谢回复！

**Comment by edwinjhlee** [comment]
> > usql 没有 相应的 module，我们 是通过一个 叫做 xbin 的机制。
> > 我承认这件事有点乱。但也没有办法。因为一个 pkg 可能有很多 bin，如果 所有这些 bin 都是 module，那么 module 数目会 暴增，而且 pkg 之前 也有 重复的 bin。
> > 你现在 可以在 x-cmd 的 搜索里面 看到 各个包 可用的方法 （ 有些 包我们还没有 来得及 介绍 ），能用 x 模式的，我们都在右上角标注了。
> 
> 嗯嗯，我就是看到有些pkg只有env use，有些却同时还有x [mod]。 等于说这种形式就是为了使得bin不重复，实际上x usql不属于真的x [mod]，是这样的吗？ 不好意思，统计学出身，学数学学的这样，会想在脑海中形成一个范式。


是的。如果是 真的 mod，那么会在 mod 目录 找到。
我们现在给这种现象一个 很粗糙的命名，叫做 xbin（ 我可能会给这个概念换一个名字，例如 轻量级模块 ） -- 这是一个容易引起混淆名字，尤其当我们后面真得在内部编写了名为 xbin 执行文件。

轻量级模块 的存在 就是 让部分 包 可以不用 安装 （影响） 环境，但用户可以 轻松调用。例如 `x usql`，本来 `x jq` 也是如此，但是后面我决定给 jq 做一些 增强，然后就有了 jq 模块。你可以看一下 https://x-cmd.com/mod/jq ， 我们做了一些 有趣的 交互增强。

我后面会给 这个 xbin，或者 轻量级模块 这个概念，起一个 准确的名字。

---

更大的一种可能，就是我后面真得 把 所有 xbin 都 模块化，因此 完全取消 这个概念。其实我一直内心 都倾向于 这种方案。

这个问题犹豫快两年了，随着 LLM 的兴起，这种方案 现在 越来越 可能获胜。

我的思路是这样，所有 值得 xbin 模式的 pkg，那么都是：
1. 这一定是工具
2.  我们推荐的，最佳的，

而这些最佳工具，都值得 x-cmd 做一些 增强 -- 尤其是 AI 的增强。

最大的问题是 mod 数目会 暴增。

---

在这个决定 定下来之前，只能 先保留这种概念的 模糊性 ～

**Comment by MaxforCherubim** [comment]
原来如此，完全理解。
后续以其他issue来精确提出问题，感谢各位的帮助，非常感谢！
其实我应该去discussions开启这个话题的，不知道能否把此issuse转移过去。

**Comment by edwinjhlee** [comment]
> 原来如此，完全理解。 后续以其他issue来精确提出问题，感谢各位的帮助，非常感谢！ 其实我应该去discussions开启这个话题的，不知道能否把此issuse转移过去。

正有此意 ～

**Comment by MaxforCherubim** [comment, marked as answer]
请查看聊天记录，相信会对X-cmd有更深的理解
